### PR TITLE
Document S3 and GCS path functionality of DataFrame.to_csv()

### LINKED ISSUE #8508: Allow writing to S3 paths
It would be really great if `to_(filetype)` supported writing to S3. 

Here is an example upload to s3 function that takes in a local file and places it on a s3 bucket.

``` python
def upload_to_s3(local_file_path, file_name, bucket_name, s3_directory):
    """
    Returns
    ----------
    Uploads local file to appropriate s3 key, and prints status 

    Parameters
    ----------
    local_file_path : str
        ex. 'my/local/path'
    file_name : str
        ex. 'cleaned_data.csv'
    bucket_name : str
        ex. 'dsapp-edu-data'
    s3_directory : str
        ex. 'NC-Cabarrus/cleaned_data'
    """

    def percent_cb(complete, total):
        """
        Helper function that prints progress
        """
        sys.stdout.write('.')
        sys.stdout.flush()

    conn = boto.connect_s3()
    bucket = conn.get_bucket(bucket_name)
    full_key_name = os.path.join(s3_directory, file_name)
    k = bucket.new_key(full_key_name)
    full_filepath = os.path.join(local_file_path, file_name)
    k.set_contents_from_filename(full_filepath, cb=percent_cb, num_cb=10)

    return None 
```

### PR DESCRIPTION
**Pull request by oguzhanogreden**
- [x] closes #8508 

I made more changes than absolutely necessary for documenting this. It's a bit more complete and it reads a more similar to `pd.read_csv()`

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -3128,10 +3128,12 @@ def to_csv(
 
         Parameters
         ----------
-        path_or_buf : str or file handle, default None
-            File path or object, if None is provided the result is returned as
-            a string.  If a file object is passed it should be opened with
-            `newline=''`, disabling universal newlines.
+        path_or_buf : str, path object, file-like object, default None
+            If a string is passed, it can be an existing local file path, as
+            well as a gcs:// or s3:// path. If a file object is passed it
```
Comment (WillAyd) [inline pandas/core/generic.py]: So I might be a little too strict, but I'm generally against documenting requirements of third party dependencies - would these not generally fall under the umbrella of being "file-like" to begin with?

**Review by WillAyd — COMMENTED** [review]
Thanks for the PR. Question on my end

**Comment by oguzhanogreden** [comment]
I agree this is actually implied. I also think that’s not helpful enough for a decent portion of pandas users who may find this useful.

I considered adding it to the user guide. But my impression is that the guidance there usually is a bit more complete.

I could always write it as an answer to a StackOverflow question and hope search indexes will help me.

**Comment by WillAyd** [comment]
Yea let’s do that. So I think you can revert the changes to pandas/core/generic, but can keep the typo fix in the IO guide and we can merge

**Comment by WillAyd** [comment]
Great thanks @oguzhanogreden

**Pull request merged**